### PR TITLE
Migrating APIs from our project repositories

### PR DESCRIPTION
**Pull request by Redjest**
Migrated the APIs from the different projects to a clean version in the platform repo. Also, removed the dependency on pytorch in the pre-processing and data loading side of the platform.

Let's make an ultra-critical review of the APIs before we merge them. Try to think of all the use cases you encountered in your work.

### CODE REVIEW

File: src/physlearn/apis/__init__.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+"""
+PhysioLearn APIs. Definition of the relations ant interfaces between different
```
Comment (galyaSegal) [inline src/physlearn/apis/__init__.py]: and*

File: src/physlearn/names.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
+from enum import Enum
+
+
```
Comment (galyaSegal) [inline src/physlearn/names.py:3]: add an issue in this repository for enums replacement

File: src/physlearn/utils.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
+from datetime import datetime, timedelta
+from hashlib import md5
```
Comment (galyaSegal) [inline src/physlearn/utils.py:2]: most of the functions here are not used (`HdfSource` uses them, and it is migrated to the repository). Need to think if we still want to have them here

File: src/physlearn/names.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    EEG = "eeg"
+    RRIntervals = "rri"
+    FAKE = "fake signal"
+    Unknown = "unknown"
+    SPO2 = "spo2"
+    HR = "hr"
+
+
+class NonSignalDataType(DataType):
+    HOSPITAL_NAME = "hospital_name"
```
Comment (galyaSegal) [inline src/physlearn/names.py:27]: 1. Some non signal data types here are specific to the epilepsy project such as `NUM_SEIZURES` 
2. Most of the non signal data types are not being used in this repository (they being used by our sources, mostly by the `HdfSource`)

**Review by galyaSegal — CHANGES_REQUESTED** [review]

File: .github/workflows/python-package-conda.yml [hunk 5]
```diff
@@ -11,7 +11,7 @@ jobs:
       - uses: actions/checkout@v2
       - name: install micromamba
         run: |
-          wget -qO- https://micromamba.snakepit.net/api/micromamba/linux-64/latest | tar -xvj bin/micromamba --strip-components=1
+          wget -qO- https://micromamba.snakepit.net/api/micromamba/linux-64/0.19.1 | tar -xvj bin/micromamba --strip-components=1
```
Comment (rotesh) [inline .github/workflows/python-package-conda.yml]: why hardcoding the version ?

File: src/physlearn/apis/data.py [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+    signal.
+      time_axis: A tensor with the time axis of the signal, shape: [1, time steps].
+      signal: The signal tensor of shape [channels, time steps].
+      signal_type: What kind of signal is it?
+      channel_names: sequence of channel names in the signal. For example: names of
+    different EEG electrodes.
+
+    """
+
+    def __init__(
```
Comment (rotesh) [inline src/physlearn/apis/data.py:24]: no longer logging sample frequency ?

File: src/physlearn/apis/data.py [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+        """ """
+        return self._end_time
+
+    @property
+    def time_axis(self):
+        """ """
+        return self._time_axis.copy()
+
+    @property
+    def signal(self):
```
Comment (rotesh) [inline src/physlearn/apis/data.py:81]: did we not want to add a magic function for easy access ?

File: src/physlearn/apis/data.py [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+    def record_id(self):
+        return self._record_id
+
+    def sample_like_this(
+        self,
+        db: Optional[DataBaseName] = None,
+        db_version: Optional[str] = None,
+        patient_id: Optional[int] = None,
+        sample_id: Optional[int] = None,
+        time: Optional[timedelta] = None,
```
Comment (rotesh) [inline src/physlearn/apis/data.py]: What's the meaning of time in the sample object ?

File: src/physlearn/apis/data.py [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+        signals: Optional[Dict[SignalType, Signal]] = None,
+        data: Optional[Dict[NonSignalDataType, Array]] = None,
+        metadata: Optional[Dict[NonSignalDataType, Any]] = None,
+        label: Dict[LabelType, Array] = None,
```
Comment (rotesh) [inline src/physlearn/apis/data.py]: I think maybe the labels should be optional too to support future unsupervised learning endeavours ?

File: src/physlearn/apis/data.py [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+    def data(self):
+        """ """
+        return self._data
+
+    @property
+    def metadata(self):
+        """ """
+        return self._metadata
+
+    def __init__(
```
Comment (rotesh) [inline src/physlearn/apis/data.py]: strictly visual, but I like __init__ to be the first method of the class

File: src/physlearn/apis/data.py [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+                self.sample_id == other.sample_id,
+                self.time == other.time,
+                self.signals == other.signals,
+                compare_tensor_dict(self.data, other.data),
+                compare_tensor_dict(self.label, other.label),
+            ]
+        )
+
+
+def compare_tensor_dict(d1: Dict[Any, Array], d2: Dict[Any, Array]):
```
Comment (rotesh) [inline src/physlearn/apis/data.py:333]: Is it still needed when working with np.arrays ?

File: src/physlearn/apis/data_loading.py [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+            - Record 2:
+                - Sample 1
+                - Sample 2
+                - ...
+            - ...
+    This allows samples to be attributed to a record which correspond to a specific
+    hospital visit in the real world.
+
+    Notes:
+        1. The term sample may mean different things in specific kinds of data, always
```
Comment (rotesh) [inline src/physlearn/apis/data_loading.py]: Why number it if there is only 1 ?

File: src/physlearn/apis/data_loading.py [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+        # for each patient
+        for patient_id in self.patient_ids:
+            # for each record
+            for record_id in self.record_ids_per_patient(patient_id):
+                # iterate over all the samples and yield their ids
+                for sample_id in self.sample_ids_per_record(patient_id, record_id):
+                    yield patient_id, record_id, sample_id
+
+
+class DataSink(ABC):
```
Comment (rotesh) [inline src/physlearn/apis/data_loading.py:224]: I still don't really get the use case of this class
and I don't think sink is explains it

File: src/physlearn/apis/labeling.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    @abstractmethod
+    def label_type(self) -> Sequence[LabelType]:
+        """
+        Returns:
+             label type of labeler.
+        """
+        pass
+
+
+class MetadataExtractor(ABC):
```
Comment (rotesh) [inline src/physlearn/apis/labeling.py:58]: What is this used for ?

File: src/physlearn/apis/metadata.py [hunk 15]
```diff
@@ -0,0 +1,45 @@
+from abc import ABC, abstractmethod
+
+from physlearn.names import DataBaseName, LabelType
+
+
+class LabelMetadata(ABC):
```
Comment (rotesh) [inline src/physlearn/apis/metadata.py:6]: Seems like an overkill to define a class for that only to be used for the labeler

File: src/physlearn/apis/processing.py [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+            allot access of all methods to all attributes of the input sample.
+            **kwargs: For concrete implementations
+
+        Returns:
+            A processed sample
+
+        """
+
+        signals = self._process_signals(x.signals, data=x.data, label=x.label)
+        data = self._process_data(x.data, signals=x.signals, label=x.label)
```
Comment (rotesh) [inline src/physlearn/apis/processing.py:36]: maybe send to data the new signals and then you can use the same processor to extract vector and scalar features ?

File: src/physlearn/apis/training/metrics.py [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+    """
+
+    def __call__(
+        self, y: Dict[LabelType, Tensor], y_pred: Dict[LabelType, Tensor]
+    ) -> float:
+        with torch.no_grad():
+            return self._metric(y, y_pred)
+
+    @abstractmethod
+    def _metric(
```
Comment (rotesh) [inline src/physlearn/apis/training/metrics.py]: I think this name is not indicative enough

File: src/physlearn/utils.py [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+      time axis: with given length, starting from start time.
+      Time difference between consecutive samples is determined by dt.
+
+    """
+    end = start + dt * length
+    time_axis = np.arange(start, end, dt).astype(datetime)
+    return time_axis
+
+
+def find_val_in_list(my_list: Sequence, value):
```
Comment (rotesh) [inline src/physlearn/utils.py]: Missing -> documentation
also, np.where / more_itertools.locate cover this functionality

File: src/physlearn/utils.py [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+    Args:
+      extension: string of desired file extension ('edf', 'txt', etc)
+      files_path: path for txt files
+
+    Returns:
+      files: a list of file paths (str) with the extension
+
+    """
+    files = []
+    for file in Path(files_path).glob("**/*." + extension):
```
Comment (rotesh) [inline src/physlearn/utils.py]: why not just return list(Path(files_path).glob(...)) ?

File: src/physlearn/utils.py [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+    Returns:
+      my_list: the given list without duplicates.
+
+    """
+    if my_list is None:
+        return my_list
+    return list(set(my_list))
+
+
+def find_elements_by_condition(my_list: list, condition: Callable):
```
Comment (rotesh) [inline src/physlearn/utils.py]: Feels redundant

File: src/physlearn/utils.py [hunk 21]
```diff
@@ -0,0 +1,266 @@
...
+      files: a list of file paths (str) with the extension
+
+    """
+    files = []
+    for file in Path(files_path).glob("**/*." + extension):
+        files.append(file)
+    return files
+
+
+def remove_duplicates_in_list(my_list: list):
```
Comment (rotesh) [inline src/physlearn/utils.py]: Feels redundant

File: src/physlearn/utils.py [hunk 22]
```diff
@@ -0,0 +1,266 @@
...
+    Check that the given path points to an hdf5 file.
+
+    Raises:
+        Value error if the file is not an hdf5 file.
+    """
+    if not h5py.is_hdf5(path):
+        raise ValueError("The given path is not an hdf5 file")
+
+
+def datetime_into_float(
```
Comment (rotesh) [inline src/physlearn/utils.py]: Also redundant ?

File: src/physlearn/utils.py [hunk 23]
```diff
@@ -0,0 +1,266 @@
...
+    """Find digits in a given string.
+
+    Args:
+      string: str, input string with the desired digits
+
+    Returns:
+      digits: str, found in the given string
+
+    """
+    digits = "".join(x for x in string if x.isdigit())
```
Comment (rotesh) [inline src/physlearn/utils.py]: I think the constant is redundant, just return

File: src/physlearn/utils.py [hunk 24]
```diff
@@ -0,0 +1,266 @@
...
+    Returns:
+      : none
+
+    """
+    msg = f"Hospital mismatch: {hospital_in} and {hospital_check.name}"
+    if hospital_in.name != hospital_check.name:
+        raise ValueError(msg)
+
+
+def bytes2str(value: Union[str, bytes]):
```
Comment (rotesh) [inline src/physlearn/utils.py]: Missing -> documentation

File: src/physlearn/apis/__init__.py (hunk 1 above)
Comment (Redjest) [reply src/physlearn/apis/__init__.py]: ok

File: src/physlearn/names.py (hunk 4 above)
Comment (Redjest) [reply src/physlearn/names.py:27]: We will migrate the sources to here too, after the Apis are in place

File: .github/workflows/python-package-conda.yml (hunk 5 above)
Comment (Redjest) [reply .github/workflows/python-package-conda.yml]: `miniconda` is under active development so they tend to release non backward compatible versions that surprisingly crash our CI

File: src/physlearn/apis/data.py (hunk 6 above)
Comment (Redjest) [reply src/physlearn/apis/data.py:24]: Signals may not have a sampling frequency (e.g. HRV signals)

File: src/physlearn/apis/data.py (hunk 7 above)
Comment (Redjest) [reply src/physlearn/apis/data.py:81]: right, added

File: src/physlearn/apis/data.py (hunk 8 above)
Comment (Redjest) [reply src/physlearn/apis/data.py]: As the init docs say

File: src/physlearn/apis/data.py (hunk 9 above)
Comment (Redjest) [reply src/physlearn/apis/data.py]: Then the user can provide an empty dict `{}`

File: src/physlearn/apis/data.py (hunk 10 above)
Comment (Redjest) [reply src/physlearn/apis/data.py]: agree

File: src/physlearn/apis/data.py (hunk 11 above)
Comment (Redjest) [reply src/physlearn/apis/data.py:333]: Yes... `array1==array2` is still not a Boolean but an array of bools

File: src/physlearn/apis/data_loading.py (hunk 12 above)
Comment (Redjest) [reply src/physlearn/apis/data_loading.py]: ok

File: src/physlearn/apis/data_loading.py (hunk 13 above)
Comment (Redjest) [reply src/physlearn/apis/data_loading.py:224]:
```python
sink[1, 4, 7] = processor(source[1, 4, 7]))
```

File: src/physlearn/apis/labeling.py (hunk 14 above)
Comment (Redjest) [reply src/physlearn/apis/labeling.py:58]: Extracts arbitrarily complicated metadata and represents it in a structured way to be used by labelers

File: src/physlearn/apis/metadata.py (hunk 15 above)
Comment (Redjest) [reply src/physlearn/apis/metadata.py:6]: The labeling is actually a pretty hard problem, even for the EFM, when we will start considering the fetal distress detection as real-time we will need good labelers

File: src/physlearn/apis/processing.py (hunk 16 above)
Comment (Redjest) [reply src/physlearn/apis/processing.py:36]: What do you think of the API we discussed?
```python
class Processor(ABC):
    """
    A base class for sample processor - a component that apply pre-processing,
    normalization, data augmentation, etc.
    """

    def __init__(self):
        if type(self, Processor):
            raise TypeError(
                "Processor API shouldn't be instantiated directly. Use subclasses."
            )

    def __call__(self, x: Sample, **kwargs) -> Sample:
        """Process a sample

        Args:
            x: The input sample to process
            **kwargs: For concrete implementations

        Returns:
            A processed sample

        """
        return self._process_sample(x)

    @abstractmethod
    def _process_sample(x: Sample) -> Sample:
        pass
```

Comment (Redjest) [reply src/physlearn/apis/processing.py:36]: I cam also implement `__matmul__` and then we can use the processors like:
```python
processed_sample = processor@raw_sample
```

File: src/physlearn/apis/training/metrics.py (hunk 17 above)
Comment (Redjest) [reply src/physlearn/apis/training/metrics.py]: Agree, changed

File: src/physlearn/utils.py (hunk 18 above)
Comment (Redjest) [reply src/physlearn/utils.py]: Agree, using bare python is clearer here

File: src/physlearn/utils.py (hunk 19 above)
Comment (Redjest) [reply src/physlearn/utils.py]: Agree

File: src/physlearn/utils.py (hunk 21 above)
Comment (Redjest) [reply src/physlearn/utils.py]: Agree

File: src/physlearn/utils.py (hunk 22 above)
Comment (Redjest) [reply src/physlearn/utils.py]: Agree, it's a single line

File: src/physlearn/utils.py (hunk 23 above)
Comment (Redjest) [reply src/physlearn/utils.py]: Right

File: src/physlearn/utils.py (hunk 24 above)
Comment (Redjest) [reply src/physlearn/utils.py]: thx

File: src/physlearn/utils.py (hunk 3 above)
Comment (Redjest) [reply src/physlearn/utils.py:2]: Removed some as they were a little too trivial, others were left to help with the migration of data sources

File: src/physlearn/apis/processing.py (hunk 16 above)
Comment (rotesh) [reply src/physlearn/apis/processing.py:36]: agreed

File: src/physlearn/apis/data.py (hunk 11 above)
Comment (rotesh) [reply src/physlearn/apis/data.py:333]: but all(arr1 == arr2) is

Comment (Redjest) [reply src/physlearn/apis/data.py:333]: Yes, but that's a dict of arrays, we need at least dict comprehension.

File: src/physlearn/utils.py (hunk 20 above)
Comment (Redjest) [reply src/physlearn/utils.py]: agree

**Review by galyaSegal — APPROVED** [review]

**Comment by galyaSegal** [comment]
why `time` was removed from Sample API?

**Comment by Redjest** [comment]
> why `time` was removed from Sample API?

All users l spoke to found it confusing, after several debates with @rotesh and our undergraduate students I got to a conclusion it's not necessary for many use cases and many people don't understand it.
